File: leetcode/go/src/leetcode/33_search_in_rotated_sorted_array.go
package leetcode

// time: O(logn), space: O(1)
func search(nums []int, target int) int {
	begin, end := 0, len(nums)-1
	if end == -1 {
		return -1
	}
	for begin < end {
		middle := (begin + end) / 2
		if nums[middle] == target {
			return middle
		}
		// middle lies in left part
		if nums[middle] >= nums[begin] {
			// target lies in left part
			if target >= nums[begin] && target < nums[middle] {
				end = middle - 1
			} else {
				begin = middle + 1
			}
		} else { // middle lies in right part
			if target > nums[middle] && target <= nums[end] {
				begin = middle + 1
			} else {
				end = middle - 1
			}
		}
	}
	if nums[begin] == target {
		return begin
	}
	return -1
}

File: leetcode/cpp/33_search_in_rotated_sorted_array.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
 public:
  int search(vector<int> &nums, int target) {
    return findInRotatedArray(nums, 0, nums.size() - 1, target);
  }

  // rotated sorted array二分查找
  int findInRotatedArray(const vector<int> &v, int left, int right,
                         int target) {
    if (left < 0 || right > static_cast<int>(v.size()) || left > right)
      return -1;
    int middle = -1;
    while (left < right) {
      middle = left + (right - left) / 2;
      if (v[middle] == target) {
        return middle;
      }
      if (v[left] <= v[middle]) {
        if (v[left] <= target && target < v[middle]) {
          right = middle - 1;
        } else {
          left = middle + 1;
        }
      } else {
        if (v[middle] < target && target <= v[right]) {
          left = middle + 1;
        } else {
          right = middle - 1;
        }
      }
    }
    return v[left] == target ? left : -1;
  }
};

int main(int argc, char *argv[]) {
  Solution s;
  vector<int> v{3, 1};
  vector<int> v1{1};
  cout << s.search(v, 1) << endl;
  cout << s.search(v1, 1) << endl;
  return 0;
}
